### ISSUE TITLE
Add a version of compile which takes options

### ISSUE THREAD

**Issue by phadej** [input-output-hk/plutus#4578]
### Describe the feature you'd like

Add a version of `compile` which take options

For example to disable optimization pass (or particular optimizations if there would be many).

After recently upgraded `plutus`, it seems that newer plutus compiler inlines too early, and therefore trumps (easy) opportunities to deduplicate bindings (cheap CSE).

**Comment by michaelpj** [comment]
There are a bunch of options you can pass with `-fplugin-opt`. They're a bit internal but you can find them here: https://github.com/input-output-hk/plutus/blob/master/plutus-tx-plugin/src/PlutusTx/Plugin.hs#L151

In particular you can turn off some optimizations.

I'd also be interested in a more specific bug report about the inliner if you have an example of where you think it's making a mistake.

**Comment by phadej** [comment]
It doesn't make invalid transformation. It just makes it too early.

A simple example is something like

```haskell
let TxOut = <3-tuple>
    Foo   = <3-tuple>
in ... TxOut a b c ... Foo x y z ...
```

Currently this is rewritten to

```haskell
... (<3-tuple> a b c) ... (<3-tuple> x y z) ...
```

but arguably

```haskell
let Tuple3 = <3-tuple>
in ... Tuple3 a b c ... Tuple3 x y z
```

may be smaller.

Let me try `no-optimize`.

**Comment by michaelpj** [comment]
Bad optimization is potentially a bug too. I'm surprised that we'd inline that. `<3-tuple>` here is a datatype definition or what?

**Comment by phadej** [comment]
It's a lambda term: `\arg1 arg2 arg3 ~ case -> case ! arg1 arg2 arg3` (~ and ! are delay and force).

**Comment by phadej** [comment]
`no-optimize` has no effect. By quickly looking at the code I see it used only once

```
plutus-tx-plugin/src/PlutusTx/Plugin.hs:                 & set (PIR.ccOpts . PIR.coOptimize) (poOptimize opts)
```

PIR is not UPLC, AFAIK. I'd prefer to have different flags for PIR and UPLC optimizations.

**Comment by phadej** [comment]
I edited the OP description, I was talking in particular about UPLC -> UPLC optimization pass.

**Comment by phadej** [comment]
My request is broadly: don't try to guess what is the right configuration, just make everything configurable, so users can tweak it.

That's a lesson learned hard-way by GHC, Cabal, ...

Helps with debugging issues too.

**Comment by effectfully** [comment]
@zliu41 do you happen to know what's the status of this issue?

**Comment by zliu41** [comment]
> what's the status of this issue?

It's been partially addressed by #4984, but we still don't support making the optimization pipeline fully configurable. Kenneth is working on the latter, and I think this issue can be closed once his work is merged.

**Comment by effectfully** [comment]
@zliu41 perfect, thank you very much. I just marked the issue as `status: needs action from the team` (where "action" = "getting Kenneth's work over the line"), so that we don't forget to follow up on it.

**Comment by effectfully** [comment]
An update: Kenneth is working on the ticket, we didn't forget about it. There was a complication with De Bruijn indices for PIR, which is an option that we wanted to have for consistency, but it turned out to be too much hassle and probably people don't really care about De Bruijn indices for PIR anyway, so we gave up on it and Kenneth is focusing on the gist of the issue.

**Comment by effectfully** [comment]
@kwxm could you please review this discussion and tell us what the current status is, if you happen to know?

**Comment by effectfully** [comment]
Asked @kwxm on slack, got the following response:

> We have a couple of issues for that (PLT-1047 and PLT-1046) but they got stalled. I did provide a way to turn optimisations on and off in the pls/pir/uplc commands (PLT-1145, PR #5189) and the intention was to come back and do the other two after that, but other things took over. It also looked as if it was going to be tricky (optimisation's kind of mixed up with compilation in general: see this function and the one below it). It might not be too hard to make it possible to turn specific optimisations on and off, but ideally we'd be able to reconfigure the entire pipeline and I think that'd need some kind of DSL to allow the user to describe what should be done: maybe you'd have to write that in a file and pass it on the command line. Maybe we should go back to this now that more optimisations have been implemented.

@michaelpj you left some comments in PLT-1047 and PLT-1046 (which aren't publicly accessible as it's our internal Jira stuff), could you please summarize your view here?

**Comment by michaelpj** [comment]
I'm somewhat positive on making the optimization infrastructure more flexible, and somewhat negative on trying to expose that all via the CLI rather than just programmatically (i.e. letting people assemble their own optimization pipeline from the pieces).

**Comment by effectfully** [comment]
It would be great to have the requested functionality, but it seems like this issue is stalled and isn't being considered a priority right now with all the other stuff going on. Well, it's a shame. Marking the issue with "Low priority".

@michaelpj if there's any ongoing or planned work related to this issue, please let us know.
